backend/msgpack/conn.go

```diff
@@ -148,22 +166,26 @@ func (c *connection) connectWithLock() error {
 	return nil
 }
 
-func (c *connection) checkReconnectWithLock() bool {
+func (c *connection) checkReconnectWithLock() error {
 	// If we haven't accumulated enough failures to warrant another reconnect
 	// and we haven't past the maximum duration since the last time we attempted
 	// to connect then we simply return false without reconnecting.
 	if c.numFailures <= c.threshold && c.nowFn().UnixNano() < c.lastConnectNanos+int64(c.maxDuration) {
-		return false
+		return errNoActiveConnection
 	}
-	if err := c.connectWithLockFn(); err == nil {
-		c.resetWithLock()
-		return true
+	err := c.resetAndConnectWithLock()
+	if err == nil {
+		c.threshold = c.initThreshold
+		return nil
 	}
 	if c.threshold < c.maxThreshold {
```

Comment:
We probably want to check whether we get here because of the failure limit or the time limit. Otherwise, we may get here because of the time limit and incorrectly bump up the failure threshold even when the number of failures is still way below the threshold.

Reply:
Good point, will update.